manager.go

```diff
@@ -0,0 +1,248 @@
+	return m.p.Store.Get(ctx, name)
+}
+
+// CreateACL creates an ACL with the given name. It also creates an ACL
+// _name which is the ACL that guards membership of the ACL itself. Any
+// member of _name or any member of the admin ACL may change the
+// membership of ACL name. Only members of the admin ACL may change the
+// membership of _name.
+//
+// The name itself must not start with an underscore.
+//
+// This does nothing if an ACL with that name already exists.
+func (h *Manager) CreateACL(ctx context.Context, name string, initialUsers ...string) error {
+	if isMetaName(name) {
+		return errgo.Newf("invalid ACL name %q", name)
+	}
+	if err := h.p.Store.CreateACL(ctx, name, initialUsers...); err != nil {
+		return errgo.Mask(err, errgo.Is(ErrBadUsername))
+	}
+	if err := h.p.Store.CreateACL(ctx, metaName(name)); err != nil {
+		return errgo.Mask(err)
+	}
+	return nil
+}
+
+// aclName is implemented by the request parameters for all endpoints
+// to return the associated ACL name.
+type aclName interface {
+	ACLName() string
+}
+
+type handler struct {
+	manager *Manager
+}
+
+// newHandler returns a handler instance to serve a particular HTTP request.
+func (m *Manager) newHandler(p httprequest.Params, arg aclName) (handler, context.Context, error) {
+	ctx := p.Context
+	if err := m.authorizeRequest(ctx, p, arg.ACLName()); err != nil {
+		return handler{}, nil, errgo.Mask(err, errgo.Any)
+	}
+	return handler{
+		manager: m,
+	}, p.Context, nil
+}
+
+// authorizeRequest checks that an HTTP request is authorized. If the
+// authorization failed because Authenticate failed, it returns an error
+// with an errAuthenticationFailed cause to signal that the desired
+// error response has already been written.
+func (m *Manager) authorizeRequest(ctx context.Context, p httprequest.Params, aclName string) error {
+	if aclName == "" {
+		return httprequest.Errorf(httprequest.CodeBadRequest, "empty ACL name")
+	}
+	identity, err := m.p.Authenticate(ctx, p.Response, p.Request)
+	if err != nil {
+		return errAuthenticationFailed
+	}
+	var checkACLName string
+	if aclName == AdminACL || isMetaName(aclName) {
+		// We're trying to access either the admin ACL or a meta-ACL; for either
+		// of these, admin privileges are needed.
+		checkACLName = AdminACL
+	} else {
+		// For all normal ACLs, access for a given ACL name is decided via membership
+		// of the meta-ACL for that name.
+		checkACLName = metaName(aclName)
+	}
+	acl, err := m.ACL(ctx, checkACLName)
+	if err != nil {
+		return errgo.Mask(err, errgo.Is(ErrACLNotFound))
+	}
+	if checkACLName != AdminACL {
+		// Admin users always get permission to do anything.
+		adminACL, err := m.ACL(ctx, AdminACL)
+		if err != nil {
+			return errgo.Notef(err, "cannot get admin ACL")
+		}
+		acl = append(acl, adminACL...)
+	}
+	ok, err := identity.Allow(ctx, acl)
+	if err != nil {
+		return errgo.Notef(err, "cannot check permissions")
+	}
+	if !ok {
+		return httprequest.Errorf(httprequest.CodeForbidden, "")
```

Comment:
I *think* this should be unauthorized rather than forbidden. My reading of the spec is that Unauthorized is for things you're not allowed to do because you don't have permission. Forbidden is for things you can't do because it breaks the laws of physics.

Reply:
This path should really provide a mechanism by which you can get authorized too.

Reply:
That's not the way I understand it. I don't seem to be the only one: https://stackoverflow.com/questions/3297048/403-forbidden-vs-401-unauthorized-http-responses

Reply:
From the RFC: 

> 10.4.4 403 Forbidden
>
> The server understood the request, but is refusing to fulfill it. Authorization will not help and the request SHOULD NOT be repeated. If the request method was not HEAD and the server wishes to make public why the request has not been fulfilled, it SHOULD describe the reason for the refusal in the entity. If the server does not wish to make this information available to the client, the status code 404 (Not Found) can be used instead. 

It's the "authorization will not help" bit that leads me to that conclusion. In this case certainly having authorization would help.

Reply:
From [RFC 7235](https://tools.ietf.org/html/rfc7235):

> A server that receives valid credentials that are not adequate to
> gain access ought to respond with the 403 (Forbidden) status code